wasm/jit/jit_amd64.go

```diff
@@ -916,6 +918,224 @@ func (c *amd64Compiler) compileMulForFloats(instruction obj.As) error {
+		c.addInstruction(jmpAtEndOfZero)
+
+		// Start emitting non-zero case.
+		// First, we calculate the most significant set bit.
+		mostSignificantSetBit := c.newProg()
+		// Set the jump target of the first non-zero check.
+		jmpIfNonZero.To.SetTarget(mostSignificantSetBit)
+		if o.Type == wazeroir.UnsignedInt32 {
+			mostSignificantSetBit.As = x86.ABSRL
+		} else {
+			mostSignificantSetBit.As = x86.ABSRQ
+		}
+		mostSignificantSetBit.From.Type = obj.TYPE_REG
+		mostSignificantSetBit.From.Reg = target.register
+		mostSignificantSetBit.To.Type = obj.TYPE_REG
+		mostSignificantSetBit.To.Reg = target.register
+		c.addInstruction(mostSignificantSetBit)
+
+		// Now we XOR the value with the bit length minus one.
+		xorWithBitLength := c.newProg()
+		xorWithBitLength.To.Type = obj.TYPE_REG
+		xorWithBitLength.To.Reg = target.register
+		xorWithBitLength.From.Type = obj.TYPE_CONST
+		if o.Type == wazeroir.UnsignedInt32 {
+			xorWithBitLength.As = x86.AXORL
+			xorWithBitLength.From.Offset = 31
+		} else {
+			xorWithBitLength.As = x86.AXORQ
+			xorWithBitLength.From.Offset = 63
+		}
+		c.addInstruction(xorWithBitLength)
+
+		// Finally the end jump instruction of zero case must target towards
+		// the next instruction.
+		c.setJmpOrigin = jmpAtEndOfZero
+	}
+
+	// We reused the same register of target for the result.
+	result := c.locationStack.pushValueOnRegister(target.register)
+	result.setRegisterType(generalPurposeRegisterTypeInt)
+	return nil
+}
+
+// compileCtz emits instructions to count up the trailing zeros in the
+// current top of the stack, and push the count result.
+// For example, stack of [..., 0xff_ff_ff_00] results in [..., 8].
+func (c *amd64Compiler) compileCtz(o *wazeroir.OperationCtz) error {
+	target := c.locationStack.pop()
+	if err := c.ensureOnGeneralPurposeRegister(target); err != nil {
+		return err
+	}
+
+	if runtime.GOOS != "darwin" {
+		countZeros := c.newProg()
+		countZeros.From.Type = obj.TYPE_REG
+		countZeros.From.Reg = target.register
+		countZeros.To.Type = obj.TYPE_REG
+		countZeros.To.Reg = target.register
+		if o.Type == wazeroir.UnsignedInt32 {
+			countZeros.As = x86.ATZCNTL
+		} else {
+			countZeros.As = x86.ATZCNTQ
+		}
+		c.addInstruction(countZeros)
+	} else {
+		// Somehow if the target value is zero, TZCNT always returs zero which is wrong,
+		// so we have the branches for non-zero and zero cases on macos.
+		// TODO: find the reference to this behavior and put the link here.
+
+		// First we compare the target with zero.
+		cmpZero := c.newProg()
+		cmpZero.As = x86.ACMPQ
+		cmpZero.From.Type = obj.TYPE_REG
+		cmpZero.From.Reg = target.register
+		cmpZero.To.Type = obj.TYPE_CONST
+		cmpZero.To.Offset = 0
+		c.addInstruction(cmpZero)
+
+		jmpIfNonZero := c.newProg()
+		jmpIfNonZero.As = x86.AJNE
+		jmpIfNonZero.To.Type = obj.TYPE_BRANCH
+		c.addInstruction(jmpIfNonZero)
+
+		// If the value is zero, we just push the const value.
+		ifZeroConst := c.newProg()
```

Comment:
can this be extracted to a function to reuse? eyes are squinting, but looks the same as above zero case.

Reply:
https://github.com/tetratelabs/wazero/pull/113/commits/9c906ab4e5290f407e5a5dc6d55291f3463f73d6 there you go!